server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentBuilder.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * A fluent builder to create {@code Iterator&lt;ToXContent&gt;} objects
+ */
+public class ChunkedToXContentBuilder implements Iterator<ToXContent> {
+
+    public static ChunkedToXContentBuilder builder(ToXContent.Params params) {
+        return new ChunkedToXContentBuilder(params);
+    }
+
+    private final ToXContent.Params params;
+    private final Stream.Builder<ToXContent> builder = Stream.builder();
+    private Iterator<ToXContent> iterator;
+
+    public ChunkedToXContentBuilder(ToXContent.Params params) {
+        this.params = params;
+    }
+
+    public ChunkedToXContentBuilder startObject() {
+        builder.add((b, p) -> b.startObject());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startObject(String name) {
+        builder.add((b, p) -> b.startObject(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder endObject() {
+        builder.add((b, p) -> b.endObject());
+        return this;
+    }
```

Comment:
Having these as separate methods seems like a smell, can we just use `.object()` everywhere?

Reply:
Could you give an example of what calls to `.object()` would look like?

Reply:
I can make the start/end methods private, yes

Reply:
Taking another look at this, removing the start/end methods makes defining objects more complicated - rather than doing
```
.startObject()
.field(...)
.array(...)
.endObject()
```
you need to do
```
.startObject(b -> b.field(...).array(...))
```
which then gets more unwieldy when you have complex inner objects, and multiple nested layers. This API is meant to be easy to use - while I understand the benefit of handling the start/end automatically, I'm not sure it's worth the extra cognitive effort needed to be forced to work with the lambdas, especially nested ones.

Reply:
I could add checks as part of creating the iterator that all objects and arrays are closed, so the failure happens close to the generation code rather than when it tries to serialize it/read it back in? It's not foolproof, but it would eliminate most sources of potential errors with the separate methods

Reply:
> while I understand the benefit of handling the start/end automatically, I'm not sure it's worth the extra cognitive effort needed to be forced to work with the lambdas, especially nested ones

I'm a big fan of things that prevent users to do wrong, but I see how this could be against the spirit of such an API.

Just thinking out loud here: would it be bad to use try-with-resources?
```
try (var ob = builder.startObject()) {
  ob.field()
    .array()
   ...
}
```
On the one end, it is not the intended usage of try -- this is a stretch, at best.
Also, it still makes things more complicated and not foolproof (one might forget to call close() -> endObject()), On the other end Intellij warns you if you do not do that.

(I was just thinking out loud, personally given the way the API is written I think explicit .startObject() / .endObject() are still the best option).

Reply:
Using `try` means breaking out of the fluent interface. Adding a API check for any unclosed objects/closing out of order shouldn't be difficult, I'll try that

Reply:
I've added check methods and appropriate tests

Reply:
Almost every object we create will have entries obtained by iterating over some collection that already exists, I think it'd be better to capture that pattern in the API here. If there's a special case that really needs to call `startObject` and `endObject` itself then it can just use `execute()`.

I don't think we need to check for correct nesting here, we do that elsewhere anyway.

Reply:
> Almost every object we create will have entries obtained by iterating over some collection that already exists

Eh sorry that's not true is it? Sometimes we spell out the entries one-by-one instead. I still think I'd prefer the nesting in the code to match the nesting in the result but I can see that others may find that too complex.

Reply:
I've done this change now - I like the elimination of start/end methods, but I'm not sure about the proliferation of nested lambdas. @ldematte what do you think?